.changeset/forty-trains-notice.md

```diff
@@ -0,0 +1,7 @@
+---
+'astro': minor
+---
+
+Makes `defineConfig` generic
+
+TypeScript will now provide feedback at the type level when using `defineConfig`. For example, it will make sure `i18n.defaultLocale` is one of the locales declared in `i18n.locales`
```

Comment:
```suggestion
Improves `defineConfig` type safety. TypeScript will now error if a group of related options do not have coherent types. For example, it will now error if `i18n.defaultLocale` is not one of the locales specified in `i18n.locales`
```